File: algorithms/leetcode/705-design-hashset/design-hashset.cpp
#include <vector>
using namespace std;

class MyHashSet {
private:
    size_t TABLE_SIZE = 93;
    vector<vector<int>> inner_set;

    size_t hash(int key) {
        return key % TABLE_SIZE;
    }
public:
    /** Initialize your data structure here. */
    MyHashSet() : inner_set{TABLE_SIZE} {

    }

    void add(int key) {
        vector<int>& bucket = inner_set[hash(key)];

        auto position = std::find(bucket.begin(), bucket.end(), key);

        if (position == bucket.end()) {
            bucket.push_back(key);
        }
    }

    void remove(int key) {
        vector<int>& bucket = inner_set[hash(key)];

        auto position = std::find(bucket.begin(), bucket.end(), key);

        if (position != bucket.end()) {
            bucket.erase(position);
        }
    }

    /** Returns true if this set contains the specified element */
    bool contains(int key) {
        vector<int>& bucket = inner_set[hash(key)];

        auto position = std::find(bucket.begin(), bucket.end(), key);

        if (position != bucket.end()) {
            return true;
        }

        return false;
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * bool param_3 = obj.contains(key);
 */

File: leetcode/705-design-hashset/design-hashset.js
const BUCKETS_SIZE = 769;

/**
 * Initialize your data structure here.
 */
const MyHashSet = function () {
  this.arr = [];
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.add = function (key) {
  const bucket = key % BUCKETS_SIZE;
  if (!this.arr[bucket]) {
    this.arr[bucket] = [];
  }

  if (this.arr[bucket].indexOf(key) === -1) {
    this.arr[bucket].push(key);
  }
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.remove = function (key) {
  const bucket = key % BUCKETS_SIZE;

  if (!Array.isArray(this.arr[bucket])) return;

  const keyIndex = this.arr[bucket].indexOf(key);

  if (keyIndex !== -1) {
    this.arr[bucket].splice(keyIndex, 1);
  }
};

/**
 * Returns true if this set contains the specified element
 * @param {number} key
 * @return {boolean}
 */
MyHashSet.prototype.contains = function (key) {
  const bucket = key % BUCKETS_SIZE;

  if (!Array.isArray(this.arr[bucket])) return false;

  return this.arr[bucket].indexOf(key) !== -1;
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * var obj = new MyHashSet()
 * obj.add(key)
 * obj.remove(key)
 * var param_3 = obj.contains(key)
 */
